dnsdist: add EDNSOptionRule

Habbie opened the pull request:
### Short description
This adds `EDNSOptionRule`, used as follows (8 is edns-client-subnet):
```
warnlog(string.format("Script starting %s", "up!"))

newServer({address="8.8.8.8", pool="g1"})
newServer({address="8.8.4.4", pool="g2"})

addAction(EDNSOptionRule(8), PoolAction("g1"))
addAction(AllRule(), PoolAction("g2"))

getServer(0):setUp()
getServer(1):setUp()
```

Needs:
- [x] tests
- [x] documentation
- [ ] a good review of the pointer mangling that I cargo culted from another function in dnsdist-ecs.cc

### Checklist
<!-- please indicate if any of these things are done/included with this Pull Request. Not all boxes need to be checked for the Pull Request to be accepted -->
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

rgacogne commented on regression-tests.dnsdist/test_Advanced.py:
```diff
@@ -1595,3 +1596,25 @@ def testTempFailureTTLBinding(self):
+    addAction(EDNSOptionRule(8), DropAction())
+    """
+
+    def testDropped(self):
```
It would be nice to add a test proving that a query without the corresponding option is not dropped :)

Habbie replied:
pushed

rgacogne commented on regression-tests.dnsdist/test_Advanced.py:
```diff
@@ -1595,3 +1596,25 @@ def testTempFailureTTLBinding(self):
+
+    _config_template = """
+    newServer{address="127.0.0.1:%s"}
+    addAction(EDNSOptionRule(8), DropAction())
```
We might want to export `EDNSOptionCode.Cookie`, `EDNSOptionCode.ECS` and so on to `Lua`.

Habbie replied:
pushed

rgacogne commented on pdns/dnsdist-ecs.cc:
```diff
@@ -433,6 +433,41 @@ int removeEDNSOptionFromOPT(char* optStart, size_t* optLen, const uint16_t optio
+bool isEDNSOptionInOpt(const char* optStart, const size_t optLen, const uint16_t optionCodeToFind)
+{
+  /* we need at least:
+   root label (1), type (2), class (2), ttl (4) + rdlen (2)*/
```
Agreed, at least dnsdist-ecs.cc might need some refactoring.

Habbie replied:
I agree, but let's do that after this PR?

rgacogne commented:
This PR now has a conflict following my merge of #6831, sorry about that :-/

Habbie commented:
> This PR now has a conflict following my merge of #6831, sorry about that :-/

Rebased and force pushed; switched to `string` like the rest of #6831 so the code is better for it!

Please re-review; I'll squash after.

Habbie commented:
Squashed!

rgacogne commented on pdns/dnsdist-lua-vars.cc:
```diff
@@ -69,6 +69,20 @@ void setupLuaVars()
     });
 
+  g_lua.writeVariable("EDNSOptionCode", std::unordered_map<string,int>{
+      {"NSID",          3 },
```
Just a nit, but perhaps this would be better:
```
{ "NSID",          EDNSOptionCode::NSID },
...
```

Habbie replied:
yes!

Habbie replied:
pushed
